### PR TITLE
DDF for SONOFF TRVZB Thermostat

### PR DESCRIPTION
**Pull request by fox34**
As discussed in #7334. Thank you @Smanar!

- Product name : TRVZB
- Manufacturer : SONOFF
- Model identifier : TRVZB

### CODE REVIEW

File: devices/sonoff/trvzb-thermostat.json [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+        {
+          "name": "attr/swversion"
+        },
+        {
+          "name": "attr/type"
+        },
+        {
+          "name": "attr/uniqueid"
+        },
+        {
```
Comment (SwoopX) [inline devices/sonoff/trvzb-thermostat.json:55]: The battery reading and checkin is missing:

```
       {
          "name": "config/battery",
          "awake": true,
          "parse": {
            "at": "0x0021",
            "cl": "0x0001",
            "ep": 1,
            "eval": "Item.val = Attr.val / 2;",
            "fn": "zcl:attr"
          },
          "read": {
            "fn": "none"
          }
        },
        {
          "name": "config/checkin",
        },
```

File: devices/sonoff/trvzb-thermostat.json [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+          "name": "config/reachable"
+        },
+        {
+          "name": "state/lastupdated"
+        },
+        {
+          "name": "state/lastupdated"
+        },
+        {
+          "name": "state/temperature"
```
Comment (SwoopX) [inline devices/sonoff/trvzb-thermostat.json:82]: The defaults for this item are tailored for temperature measurement sensors and hence, will not work here. Please replace it by the following:

```
          "name": "state/temperature",
          "refresh.interval": 3660,
          "parse": {
            "at": "0x0000",
            "cl": "0x0201",
            "ep": 1,
            "eval": "Item.val = Attr.val;",
            "fn": "zcl:attr"
          },
          "read": {
            "at": "0x0000",
            "cl": "0x0201",
            "ep": 1,
            "fn": "zcl:attr"
          }
```

File: devices/sonoff/trvzb-thermostat.json [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+        }
+      ]
+    }
+  ],
+  "bindings": [
```
Comment (SwoopX) [inline devices/sonoff/trvzb-thermostat.json:87]: The battery binding is missing:

```
    {
      "bind": "unicast",
      "src.ep": 1,
      "dst.ep": 1,
      "cl": "0x0001",
      "report": [
        {
          "at": "0x0021",
          "dt": "0x20",
          "min": 3600,
          "max": 43200,
          "change": "0x00000002"
        }
      ]
    },
```

File: devices/sonoff/trvzb-thermostat.json [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+      "src.ep": 1,
+      "cl": "0x0201",
+      "report": [
+        {
+          "at": "0x0000",
+          "dt": "0x29",
+          "min": 30,
+          "max": 360,
+          "change": "0x0000000A"
+        }
```
Comment (SwoopX) [inline devices/sonoff/trvzb-thermostat.json:99]: Please add additional thermostat bindings

```
{
          "at": "0x0012",
          "dt": "0x29",
          "min": 1,
          "max": 3600,
          "change": "0x00000001"
        },
        {
          "at": "0x001C",
          "dt": "0x30",
          "min": 1,
          "max": 3600
        }
```

File: devices/sonoff/trvzb-thermostat.json [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+      "bind": "unicast",
+      "src.ep": 1,
+      "cl": "0x0201",
+      "report": [
+        {
+          "at": "0x0000",
+          "dt": "0x29",
+          "min": 30,
+          "max": 360,
+          "change": "0x0000000A"
```
Comment (SwoopX) [inline devices/sonoff/trvzb-thermostat.json:98]: Please change the values to:

```
          "min": 60,
          "max": 300,
          "change": "0x00000064"
```

File: devices/sonoff/trvzb-thermostat.json [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+        {
+          "name": "config/offset",
+          "default": 0
+        },
+        {
+          "name": "config/on"
+        },
+        {
+          "name": "state/on"
+        },
```
Comment (SwoopX) [inline devices/sonoff/trvzb-thermostat.json:71]: This item is not supported, so please remove.

File: devices/sonoff/trvzb-thermostat.json [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+        {
+          "name": "attr/uniqueid"
+        },
+        {
+          "name": "config/heatsetpoint",
+          "default": 0
+        },
+        {
+          "name": "config/mode"
+        },
```
Comment (SwoopX) [inline devices/sonoff/trvzb-thermostat.json:61]: Please investigate which modes are supported by the device. The code below must be amended according to the findings:

```
        {
          "name": "config/mode",
          "refresh.interval": 3660,
          "read": {
            "at": "0x001C",
            "cl": "0x0201",
            "ep": 1,
            "fn": "zcl:attr"
          },
          "parse": {
            "at": "0x001C",
            "cl": "0x0201",
            "ep": 1,
            "eval": "if (Attr.val == 0) { Item.val = 'off' } else if (Attr.val == 1) { Item.val = 'auto' } else if (Attr.val == 4) { Item.val = 'heat' };",
            "fn": "zcl:attr"
          },
          "write": {
            "at": "0x001C",
            "cl": "0x0201",
            "dt": "0x10",
            "ep": 1,
            "eval": "if (Item.val == 'off') { 0 } else if (Item.val == 'auto') { 1 } else if (Item.val == 'heat') { 4 };",
            "fn": "zcl:attr"
          },
          "default": "heat"
        },
```

File: devices/sonoff/trvzb-thermostat.json [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+      ],
+      "fingerprint": {
+        "profile": "0x0104",
+        "device": "0x0301",
+        "endpoint": "0x01",
+        "in": [
+          "0x0000",
+          "0x0201"
+        ]
+      },
```
Comment (SwoopX) [inline devices/sonoff/trvzb-thermostat.json:26]: In order to make use of the `config/mode` item, the following code must be added and contain the actually usable settings (see respetive item comment)

```
      "meta": {
        "values": {
          "config/mode": {"off": 0, "auto": 1, "heat": 4}
        }
      },
```

File: devices/sonoff/trvzb-thermostat.json [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+        },
+        {
+          "name": "attr/type"
+        },
+        {
+          "name": "attr/uniqueid"
+        },
+        {
+          "name": "config/heatsetpoint",
+          "default": 0
```
Comment (SwoopX) [inline devices/sonoff/trvzb-thermostat.json:57]: Please remove the default value and add `"refresh.interval": 3660` instead.

File: devices/sonoff/trvzb-thermostat.json [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+        },
+        {
+          "name": "config/mode"
+        },
+        {
+          "name": "config/offset",
+          "default": 0
+        },
```
Comment (SwoopX) [inline devices/sonoff/trvzb-thermostat.json:65]: Please replace by the code below:

```
        {
          "name": "config/offset",
          "refresh.interval": 3660,
          "read": {
            "at": "0x0010",
            "cl": "0x0201",
            "ep": 1,
            "fn": "zcl:attr"
          },
          "parse": {
            "at": "0x0010",
            "cl": "0x0201",
            "ep": 1,
            "eval": "Item.val = Attr.val * 10;",
            "fn": "zcl:attr"
          },
          "write": {
            "at": "0x0010",
            "cl": "0x0201",
            "dt": "0x28",
            "ep": 1,
            "eval": "Item.val / 10;",
            "fn": "zcl:attr"
          },
          "range": [-250, 250],
          "default": 0
        },
```

**Review by SwoopX — CHANGES_REQUESTED** [review]
Hi there,

I'm afraid, the DDF requires some significant changes to be considered. The suggested file would be barely functional.

**Comment by Smanar** [comment]
@fox34 are you still here ?

**Comment by fox34** [comment]
I made the requested changes, however:

> Please investigate which modes are supported by the device.

How would I do this? The device can be toggled between automatic and manual operation (and turned off, obviously). Automatic mode can (only?) be configured in the "official" SONOFF app, which I cannot use because I do not own the SONOFF gateway.

**Comment by Smanar** [comment]
>How would I do this? The device can be toggled between automatic and manual operation (and turned off, obviously). Automatic mode can (only?) be configured in the "official" SONOFF app, which I cannot use because I do not own the SONOFF gateway.

Yes, it's fine, just check wich one mode you can use on it.
From Z2M they are using same 

`.withSystemMode(['off', 'auto', 'heat'], ea.ALL, 'Mode of the thermostat')`

And they have too problem with the "auto" mode, it seem this device don't use zigbee classic request, and use instead a personnal one > https://github.com/Koenkk/zigbee-herdsman-converters/pull/6443

**Review by SwoopX — APPROVED** [review]
Looks ok for me wrt the requested changes, thanks!

**Comment by manup** [comment]
So ready for merging? :)

**Comment by Smanar** [comment]
For me yes, have checked again ,all @SwoopX remark are resolved.

**Pull request merged**